prow/cmd/deck/rerun.go

```diff
@@ -18,21 +18,159 @@ package main
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"net/http"
+	"strings"
 
 	"github.com/sirupsen/logrus"
 	kerrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
 	prowapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	v1 "k8s.io/test-infra/prow/apis/prowjobs/v1"
 	prowv1 "k8s.io/test-infra/prow/client/clientset/versioned/typed/prowjobs/v1"
 	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/gerrit/client"
+	"k8s.io/test-infra/prow/github"
 	"k8s.io/test-infra/prow/githuboauth"
 	"k8s.io/test-infra/prow/pjutil"
 	"k8s.io/test-infra/prow/plugins"
 	"k8s.io/test-infra/prow/plugins/trigger"
 )
 
+var (
+	// Stores the annotations and labels that are generated
+	// and specified within components.
+	ComponentSpecifiedAnnotationsAndLabels = sets.NewString(
+		// Labels
+		client.GerritRevision,
+		client.GerritPatchset,
+		client.GerritReportLabel,
+		github.EventGUID,
+		"created-by-tide",
+		// Annotations
+		client.GerritID,
+		client.GerritInstance,
+	)
+)
+
+func verifyRerunRefs(refs *prowapi.Refs) error {
+	if refs == nil {
+		return errors.New("Refs must be supplied")
+	}
+	if len(refs.Org) == 0 {
+		return errors.New("org must be supplied")
+	}
+	if len(refs.Repo) == 0 {
+		return errors.New("repo must be supplied")
+	}
+	if len(refs.BaseRef) == 0 {
+		return errors.New("baseRef must be supplied")
+	}
+	return nil
+}
+
+func setRerunOrgRepo(refs *prowapi.Refs, labels map[string]string) string {
+	org, repo := refs.Org, refs.Repo
+	orgRepo := org + "/" + repo
+	// Add "https://" prefix to orgRepo if this is a gerrit job.
+	// (Unfortunately gerrit jobs use the full repo URL as the identifier.)
+	prefix := "https://"
+	if labels[client.GerritRevision] != "" && !strings.HasPrefix(orgRepo, prefix) {
+		orgRepo = prefix + orgRepo
+	}
+	return orgRepo
+}
+
+func getPresubmitSpec(cfg config.Getter, name string, refs *prowapi.Refs, labels map[string]string) (*v1.ProwJobSpec, map[string]string, map[string]string, error) {
+	// presubmit jobs require Refs
+	if err := verifyRerunRefs(refs); err != nil {
+		return nil, nil, nil, err
+	}
+	var presubmitJob *config.Presubmit
+	branch := refs.BaseRef
+	orgRepo := setRerunOrgRepo(refs, labels)
+	// Get presubmits from Config alone.
+	presubmits := cfg().GetPresubmitsStatic(orgRepo)
+	for _, job := range presubmits {
+		job := job
+		if !job.CouldRun(branch) { // filter out jobs that are not branch matching
+			continue
+		}
+		if job.Name == name {
+			if presubmitJob != nil {
+				return nil, nil, nil, fmt.Errorf("%s matches multiple prow jobs from orgRepo %q", name, orgRepo)
+			}
+			presubmitJob = &job
```

Comment:
Why not `break` here? Also avoids the need to copy the `job` variable in the loop context

Reply:
ah, its because there might be multiple jobs with the same name. Hm.

Reply:
that's correct